### PR TITLE
Addition of download handlers

### PR DESCRIPTION
**Pull request by allroundexperts**
Added download handlers for iOS (wkwebview only) and Android (API 21+).

### CODE REVIEW

**Review by alex-english-desiclabs — APPROVED** [review]

File: www/inappbrowser.js [hunk 1]
```diff
@@ -17,109 +17,136 @@
...
-            } else if (injectDetails.file) {
-                exec(cb, null, 'InAppBrowser', 'injectStyleFile', [injectDetails.file, !!cb]);
-            } else {
-                throw new Error('insertCSS requires exactly one of code or file to be specified');
-            }
-        }
+  function InAppBrowser() {
+    this.channels = {
+      beforeload: channel.create("beforeload"),
+      ondownload: channel.create("ondownload"),
```
Comment (stevenroussey-privicy) [inline www/inappbrowser.js]: Despite all the whitespace and quote changes, this is the only real change, right? Adding `ondownload`? Is that the start of a download, or the finish of a download? I know in Electron, there is start, end, progress, and error.

**Comment by allroundexperts** [comment]
That really depends on the kind of response I send from the native side.

File: src/ios/CDVWKInAppBrowser.m [hunk 2]
```diff
@@ -1178,6 +1178,28 @@ - (void)webView:(WKWebView *)theWebView didStartProvisionalNavigation:(WKNavigat
     return [self.navigationDelegate didStartProvisionalNavigation:theWebView];
 }
 
+- (void)webView:(WKWebView *)theWebView decidePolicyForNavigationResponse:(nonnull WKNavigationResponse *)navigationResponse decisionHandler:(nonnull void (^)(WKNavigationResponsePolicy))decisionHandler
+{
+    NSArray* downloadTypes = @[@"zip", @"rar", @"tar", @"binary"];
```
Comment (stevenroussey-privicy) [inline src/ios/CDVWKInAppBrowser.m]: might add `gz` as tar files are usually `tar.gz`

File: www/inappbrowser.js (hunk 1 above)
Comment (allroundexperts) [reply www/inappbrowser.js]: That really depends on the kind of response I send from the native side.

File: src/ios/CDVWKInAppBrowser.m (hunk 2 above)
Comment (allroundexperts) [reply src/ios/CDVWKInAppBrowser.m]: Sure!

Comment (allroundexperts) [reply src/ios/CDVWKInAppBrowser.m]: Would work anyway.

File: src/android/InAppBrowser.java [hunk 3]
```diff
@@ -1362,11 +1364,59 @@ public WebResourceResponse shouldInterceptRequest (final WebView view, String ur
...
-            return response;
+        /**
+         * Sets custom download handler.
+         *
+         * @param response
+         * @return boolean
+         */
+
+        private boolean setCustomDownloadHandler(HttpURLConnection response){
+            String[] downloadExt = {"zip", "binary", "tar", "rar"};
```
Comment (stevenroussey-privicy) [inline src/android/InAppBrowser.java]: I would think we want to look for the Content-Disposition header above all else.

Comment (stevenroussey-privicy) [reply src/android/InAppBrowser.java]: It specifically tells us that the url is a download, and optionally what the filename is.

Comment (allroundexperts) [reply src/android/InAppBrowser.java]: `Content-Disposition` is an optional header. We can be fully sure that it is a download, if we use content type.

File: src/ios/CDVWKInAppBrowser.m [hunk 4]
```diff
@@ -1180,26 +1180,40 @@ - (void)webView:(WKWebView *)theWebView didStartProvisionalNavigation:(WKNavigat
...
-        }];
-        [dataTask resume];
-        decisionHandler(WKNavigationActionPolicyCancel);
+    if(index == NSNotFound) {
+        decisionHandler(WKNavigationResponsePolicyAllow);
+        return;
+    };
+    NSURLSession *session = [NSURLSession sharedSession];
+    NSURLSessionDataTask *dataTask = [session dataTaskWithURL:theWebView.URL completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
+        NSString* savePath = [self getDownloadSavePath: response.suggestedFilename];
```
Comment (stevenroussey-privicy) [inline src/ios/CDVWKInAppBrowser.m:1196]: what if there is no suggested filename?

**Comment by allroundexperts** [comment]
Not possible. It generates a default name if there is no
`Content-Disposition` header.
https://developer.apple.com/documentation/foundation/nsurlresponse/1415924-suggestedfilename?language=objc

File: src/ios/CDVWKInAppBrowser.m (hunk 4 above)
Comment (allroundexperts) [reply src/ios/CDVWKInAppBrowser.m:1196]: Not possible. It generates a default name if there is no
`Content-Disposition` header.
https://developer.apple.com/documentation/foundation/nsurlresponse/1415924-suggestedfilename?language=objc

**Pull request merged**